src/esp/io/URDFParser.h

```diff
@@ -0,0 +1,314 @@
+  URDF_CONTACT_HAS_INERTIA_SCALING = 2,
+  URDF_CONTACT_HAS_CONTACT_CFM = 4,
+  URDF_CONTACT_HAS_CONTACT_ERP = 8,
+  URDF_CONTACT_HAS_STIFFNESS_DAMPING = 16,
+  URDF_CONTACT_HAS_ROLLING_FRICTION = 32,
+  URDF_CONTACT_HAS_SPINNING_FRICTION = 64,
+  URDF_CONTACT_HAS_RESTITUTION = 128,
+  URDF_CONTACT_HAS_FRICTION_ANCHOR = 256,
+};
+
+struct URDFLinkContactInfo {
+  float m_lateralFriction;
+  float m_rollingFriction;
+  float m_spinningFriction;
+  float m_restitution;
+  float m_inertiaScaling;
+  float m_contactCfm;
+  float m_contactErp;
+  float m_contactStiffness;
+  float m_contactDamping;
+
+  int m_flags;
+
+  URDFLinkContactInfo()
+      : m_lateralFriction(0.5),
+        m_rollingFriction(0),
+        m_spinningFriction(0),
+        m_restitution(0),
+        m_inertiaScaling(1),
+        m_contactCfm(0),
+        m_contactErp(0),
+        m_contactStiffness(1e4),
+        m_contactDamping(1) {
+    m_flags = URDF_CONTACT_HAS_LATERAL_FRICTION;
+  }
+};
+
+struct UrdfLink {
+  std::string m_name;
+  UrdfInertia m_inertia;
+  Magnum::Matrix4 m_linkTransformInWorld;
+  std::vector<UrdfVisual> m_visualArray;
+  std::vector<UrdfCollision> m_collisionArray;
+  std::shared_ptr<UrdfLink> m_parentLink;
+  std::shared_ptr<UrdfJoint> m_parentJoint;
+
+  std::vector<std::shared_ptr<UrdfJoint>> m_childJoints;
+  std::vector<std::shared_ptr<UrdfLink>> m_childLinks;
+
+  int m_linkIndex;
+
+  URDFLinkContactInfo m_contactInfo;
+
+  UrdfLink() : m_parentLink(0), m_parentJoint(0), m_linkIndex(-2) {}
+};
+
+struct UrdfModel {
+  std::string m_name;
+  std::string m_sourceFile;
+  Magnum::Matrix4 m_rootTransformInWorld(Magnum::Math::IdentityInitT);
+
+  //! map of names to materials
+  std::map<std::string, std::shared_ptr<UrdfMaterial>> m_materials;
+
+  //! map of names to links
+  std::map<std::string, std::shared_ptr<UrdfLink>> m_links;
+
+  //! map of link indices to names
+  std::map<int, std::string> m_linkIndicesToNames;
+
+  //! map of names to joints
+  std::map<std::string, std::shared_ptr<UrdfJoint>> m_joints;
+
+  //! list of root links (usually 1)
+  std::vector<std::shared_ptr<UrdfLink>> m_rootLinks;
+  bool m_overrideFixedBase;
+
+  void printKinematicChain() const;
+
+  std::shared_ptr<UrdfLink> getLink(std::string linkName) const {
+    if (m_links.count(linkName)) {
+      return m_links.at(linkName);
+    }
+    return nullptr;
+  }
+  std::shared_ptr<UrdfLink> getLink(int linkIndex) const {
+    if (m_linkIndicesToNames.count(linkIndex)) {
+      return getLink(m_linkIndicesToNames.at(linkIndex));
+    }
+    return nullptr;
+  }
+
+  //! get the parent joint of a link
+  std::shared_ptr<UrdfJoint> getJoint(int linkIndex) const {
+    if (m_linkIndicesToNames.count(linkIndex)) {
+      return getLink(m_linkIndicesToNames.at(linkIndex))->m_parentJoint;
+    }
+    return nullptr;
+  }
+
+  UrdfModel() : m_overrideFixedBase(false) {}
+};
+
+class URDFParser {
+  // datastructures
+  UrdfModel m_urdfModel;
```

Comment:
It seems weird that URDFParser owns the model object. This means the parser must persist for as long as you want to use the object. I would expect to throw away a parser object after parsing is finished.

Why not just a static parse method that returns the model?